homeassistant/components/juicenet/__init__.py

```diff
@@ -1,68 +1,88 @@
-"""Support for Juicenet cloud."""
+"""The JuiceNet integration."""
+import asyncio
 import logging
 
-import pyjuicenet
+from pyjuicenet import Api
 import voluptuous as vol
 
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
 from homeassistant.const import CONF_ACCESS_TOKEN
-from homeassistant.helpers import discovery
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.entity import Entity
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv
+
+from .const import DOMAIN, EXCEPTIONS
+from .device import JuiceNetApi
 
 _LOGGER = logging.getLogger(__name__)
 
-DOMAIN = "juicenet"
+PLATFORMS = ["sensor", "switch"]
 
 CONFIG_SCHEMA = vol.Schema(
     {DOMAIN: vol.Schema({vol.Required(CONF_ACCESS_TOKEN): cv.string})},
     extra=vol.ALLOW_EXTRA,
 )
 
-JUICENET_COMPONENTS = ["sensor", "switch"]
 
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the JuiceNet component."""
+    conf = config.get(DOMAIN)
+    hass.data.setdefault(DOMAIN, {})
 
-def setup(hass, config):
-    """Set up the Juicenet component."""
-    hass.data[DOMAIN] = {}
+    if not conf:
+        return True
 
-    access_token = config[DOMAIN].get(CONF_ACCESS_TOKEN)
-    hass.data[DOMAIN]["api"] = pyjuicenet.Api(access_token)
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+        )
+    )
+    return True
 
-    for component in JUICENET_COMPONENTS:
-        discovery.load_platform(hass, component, DOMAIN, {}, config)
 
-    return True
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up JuiceNet from a config entry."""
 
+    config = entry.data
 
-class JuicenetDevice(Entity):
-    """Represent a base Juicenet device."""
+    # Configure API
+    access_token = config[CONF_ACCESS_TOKEN]
+    api = Api(access_token)
 
-    def __init__(self, device, sensor_type, hass):
-        """Initialise the sensor."""
-        self.hass = hass
-        self.device = device
-        self.type = sensor_type
+    juicenet = JuiceNetApi(api, hass)
 
-    @property
-    def name(self):
-        """Return the name of the device."""
-        return self.device.name()
+    try:
+        await hass.async_add_executor_job(juicenet.setup, hass)
+    except EXCEPTIONS as error:
```

Comment:
If this throws `ConfigEntryNotReady ` because of invalid auth that won't be recoverable and Home Assistant is going to forever try again to setup the config entry.  Please trap invalid auth separately and return instead

Reply:
Thanks for pointing this out.